### PR TITLE
[Dashboard] Total, total in and total out values not updating

### LINKED ISSUE #3359: [Dashboard] Total, total in and total out values not updating
### Issue

Payment updates in the UI required reloading the dashboard to see all the totals in the finance cards and total in and out widget to update. This is confusing because the recent activity updates, but everything else doesn't.

### PR DESCRIPTION
**Pull request by mmioana**
## Description

- This PR addresses the following issue: **Payment updates in the UI required reloading the dashboard to see all the totals in the finance cards and total in and out widget to update. This is confusing because the recent activity updates, but everything else doesn't.**

## Testing

TODO: Now it's time to test payments and transfer funds actions and motions and their effect on the `Total`, `Payments` and `Income` sections, so please bring some patience 😄 

* Step 1. Make sure you have some funds in `Andromeda` and check the initial values for the `Total`, `Payments` and `Income` sections

Screenshot 2024-10-18 at 10 35 39

* Step 2. Let's make a `Simple payment` with `Permissions`

Screenshot 2024-10-18 at 10 36 06

* Step 3. Given an action is immediately executed, the values should refresh 

Screenshot 2024-10-18 at 10 36 14

* Step 4. Create an `Advanced payment` with `Permissions` and use `0` as claim delay

Screenshot 2024-10-18 at 10 36 43

* Step 5. Execute all the steps and once you get to the `Payment` step the values in the cards should update
Screenshot 2024-10-18 at 10 37 09
Screenshot 2024-10-18 at 10 37 14

* Step 6. Let's also transfer some funds between `Andromeda` and `Serenity` - would be nice if you switch to `Serenity` to check the initial value, which clearly I forgot to screenshot

Screenshot 2024-10-18 at 10 37 53

* Step 7. As the action is immediately executed, you should see these values updating

Screenshot 2024-10-18 at 10 37 58
Screenshot 2024-10-18 at 10 38 04

* Step 8. Now let's install `Multi-sig` extension

Screenshot 2024-10-18 at 10 38 38

* Step 9. Give `leela` `Owner` permissions for `Takes actions via Multi-Sig` in `General` - thus these will be inherited also for `Andromeda`

Screenshot 2024-10-18 at 10 39 36

* Step 10. Time to create a `Simple payment` using `Multi-Sig` from `Andromeda`

Screenshot 2024-10-18 at 10 40 09

* Step 11. Finalise the `Multi-Sig` motion
 
Screenshot 2024-10-18 at 10 40 21

* Step 12. Check the values for `Total`, `Payments` and `Income`. **Only `Total` should have refreshed** - this is because the motion executed date will take into account the network time which on `dev` is not in sync with the current time

Screenshot 2024-10-18 at 10 40 56

* Step 13. Let's install the `Reputation weighted` extension

Screenshot 2024-10-18 at 10 41 24

* Step 14. Create a `Simple payment` with `Reputation` and fully support it

Screenshot 2024-10-18 at 10 42 04

* Step 15. Finalise the `Reputation` motion

Screenshot 2024-10-18 at 10 47 59

* Step 16. Check the values for `Total`, `Payments` and `Income`. **Once again only `Total` should have refreshed** - this is because the motion executed date will take into account the network time which on `dev` is not in sync with the current time

Screenshot 2024-10-18 at 10 48 17

* Step 17. Now let's switch between the domains to check the values corresponding to that domain are shown and not a value fetched for another domain - why we need to check this is because previously once switching through the domains fired multiple `graphql` requests but while you would be on `Andromeda` for eg. a value previously fetched for `General` would resolve, thus displaying the values for `General`. This issue was solved through the usage of the `AbortController`, however there were some issues for keeping the instance in sync with the request for which a cancellation should be made.

https://github.com/user-attachments/assets/cf76f6bc-d7ea-43dd-bed1-35d25b3a1be6

## Further testing

1. Consider also adding a claim delay for an `Advanced payment`. You can use `forward-time` and do a page refresh, but keep in mind to refresh the page only before reaching the `Payment` step, otherwise you'll trigger a refetch of the values and not see them being updated automatically.
2. Please try to reject a motion. This should have no impact on the `Total`, `Payments` and `Income` sections.

For other testing scenarios, let your imagination run wild 🌈 

## Diffs

**Changes** 🏗

* Add `notifyOnNetworkStatusChange` for `useGetDomainBalanceQuery` query options

Resolves #3359

### CODE REVIEW

**Review by rumzledz — APPROVED** [review]
I followed all the test steps and all is well! Thanks @mmioana for the patience while I was testing this as well 🙌 

## Simple payment

**Before**
Screenshot 2024-10-18 at 18 00 55

**After**
Screenshot 2024-10-18 at 18 01 29

## Advanced payment

**Before**
Screenshot 2024-10-18 at 18 02 07

**After**
Screenshot 2024-10-18 at 18 02 49

## Reputation

rep

## Multi-sig

MS

**Review by rdig — APPROVED** [review]
Tested all steps and the value is updating correctly each time. 

Nicely done.

Screenshot from 2024-10-21 13-55-53
Screenshot from 2024-10-21 13-59-21
Screenshot from 2024-10-21 13-59-32
Screenshot from 2024-10-21 14-00-25
Screenshot from 2024-10-21 14-00-27
Screenshot from 2024-10-21 14-01-15
Screenshot from 2024-10-21 14-01-18
Screenshot from 2024-10-21 14-01-42
Screenshot from 2024-10-21 14-02-44
Screenshot from 2024-10-21 14-03-55
Screenshot from 2024-10-21 14-03-57
Screenshot from 2024-10-21 14-05-24
Screenshot from 2024-10-21 14-05-26

**Review by bassgeta — CHANGES_REQUESTED** [review]
What a chonker to test :sweat_smile: But a graet job untangling all of these edge cases, it was probably a hair pulling experience!
1. Andromeda in and out :heavy_check_mark: (i have 2 actions because I took initial screenshots with CLNY as the conversion currency and it was confusing)
image
2 and 3 Starts loading and updates immediately :heavy_check_mark: 
image and sweet sweet 11 EUR more in "Out"
image
4 and 5. Advanced payments refresh the page (cc @mmioana I got this when creating an advanced payment, should we do this only after the last step? :eyes: )
Loading after release :heavy_check_mark: 
image
image

6. Serenity initial state :eyes: 
image
image
image
And finally the income changes :heavy_check_mark: 
image

10. Multisig created :heavy_check_mark: 
image
image

16. With voting reputation, stuff gets indeed reloaded!
image
Before finalization:
image
Afterwards, the total changes :heavy_check_mark: 
image

However I have sadly found 2 bugs :')
1. Whenever I open a multisig motion or a voting reputation motion, stuff starts reloading :thinking: )
https://github.com/user-attachments/assets/40e554e4-b86e-481b-b45d-67511cf08742

2.  I've been able to constistently get the `All teams` funds card stuck in a loading state. I can't tell you how exactly to reproduce it, it  just happened a lot of times for me. maybe it's a non issue
image

File: src/components/v5/frame/ColonyHome/partials/FundsCards/hooks.ts [hunk 1]
```diff
@@ -53,13 +54,27 @@ export const useTotalData = (domainId?: string) => {
     memoizedQueryVariables.queryOptions,
   );
 
+  const cancelQuery = useCallback(() => {
+    if (loading) {
+      memoizedQueryVariables.abortController.abort();
+    }
+  }, [memoizedQueryVariables.abortController, loading]);
+
   useEffect(() => {
```
[the remark this answers is not in the document]
Comment (mmioana) [reply src/components/v5/frame/ColonyHome/partials/FundsCards/hooks.ts:63]: No, you're right @bassgeta, for the `FundsCards` only one should be enough 🙌

[the remark this answers is not in the document]
Comment (mmioana) [reply src/components/v5/frame/ColonyHome/partials/FundsCards/hooks.ts:63]: And actually I think this might have caused the infinite loading for `All teams` funds card

**Comment by rdig** [comment]
@bassgeta @mmioana 

> I've been able to constistently get the All teams funds card stuck in a loading state. I can't tell you how exactly to reproduce it, it just happened a lot of times for me. maybe it's a non issue

I saw that as well, and it usually took a fool page reload for data to start showing up

**Comment by mmioana** [comment]
> @bassgeta @mmioana
> 
> > I've been able to constistently get the All teams funds card stuck in a loading state. I can't tell you how exactly to reproduce it, it just happened a lot of times for me. maybe it's a non issue
> 
> I saw that as well, and it usually took a fool page reload for data to start showing up

@rdig can you please re-test this as I pushed a change yesterday which I believe will fix this issue, otherwise would suggest opening a new issue for this and merge this PR

**Review by rdig — CHANGES_REQUESTED** [review]
I can still replicate the Total loading issue, fairly easily


https://github.com/user-attachments/assets/69263c2a-071d-40c3-bda4-ce7de43245e2

File: src/components/v5/common/ActionSidebar/partials/ActionSidebarContent/ActionSidebarContent.tsx [hunk 2]
```diff
@@ -316,16 +314,13 @@ const ActionSidebarContent: FC<ActionSidebarContentProps> = ({
...
+             * By default it will refetch all active queries
              */
-            client.cache.evict({
-              fieldName: 'getDomainBalance',
-            });
-            if (isQueryActive('GetDomainBalanceQuery')) {
-              client.refetchQueries({
-                include: [GetDomainBalanceDocument],
-              });
+            if (!isMotion && !isMultiSig) {
```
Comment (bassgeta) [inline src/components/v5/common/ActionSidebar/partials/ActionSidebarContent/ActionSidebarContent.tsx:322]: So since we do this for motion and multisig finalization, what if we remove the cache entry just for payment actions? so maybe we can do it in the sagas or in the forms like this?
image

File: src/components/v5/common/ActionSidebar/partials/Motions/Motions.tsx [hunk 3]
```diff
@@ -79,6 +82,17 @@ const Motions: FC<MotionsProps> = ({ transactionId }) => {
     stopPollingForAction,
   ]);
 
+  useEffect(() => {
+    // previousIsFinalized can be undefined when opening an already closed action
+    if (motionData?.isFinalized && previousIsFinalized === false) {
+      /**
+       * We need to remove all getDomainBalance queries once a payment or funding has been successfully completed
```
Comment (bassgeta) [inline src/components/v5/common/ActionSidebar/partials/Motions/Motions.tsx:89]: We currently do this for all finalized motions right? do we need to check for the action type and maybe add it to  the `useFinalizeSuccessCallback` helper hook?
image

File: src/components/v5/common/ActionSidebar/partials/MultiSigSidebar/partials/MultiSigWidget/partials/FinalizeStep/FinalizeStep.tsx [hunk 4]
```diff
@@ -221,6 +224,16 @@ const FinalizeStep: FC<FinalizeStepProps> = ({
     }
   }, [isMultiSigExecuted, isMultiSigRejected, action]);
 
+  useEffect(() => {
+    if (isMultiSigExecuted && previousIsMultiSigExecuted === false) {
+      /**
+       * We need to remove all getDomainBalance queries once a payment or funding has been successfully completed
```
Comment (bassgeta) [inline src/components/v5/common/ActionSidebar/partials/MultiSigSidebar/partials/MultiSigWidget/partials/FinalizeStep/FinalizeStep.tsx:230]: Same here as with motions, should we add this check to the `useFinalizeSuccessCallback` hook and check for the correct action type so we don't remove the cache for each finalized multisig?
image

File: src/context/TotalInOutBalanceChartContext/TotalInOutBalanceChartContextProvider.tsx [hunk 5]
```diff
@@ -62,13 +65,27 @@ const TotalInOutBalanceChartContextProvider: FC<PropsWithChildren> = ({
     memoizedQueryVariables.queryOptions,
   );
 
+  const cancelQuery = useCallback(() => {
+    if (loading) {
+      memoizedQueryVariables.abortController.abort();
+    }
+  }, [memoizedQueryVariables.abortController, loading]);
+
   useEffect(() => {
```
Comment (bassgeta) [inline src/context/TotalInOutBalanceChartContext/TotalInOutBalanceChartContextProvider.tsx:74]: do we still need both useEffects since the hook for the funding cards only does it in one `useEffect`? Maybe we could try to just have one and see if it works :crossed_fingers:

**Review by bassgeta — CHANGES_REQUESTED** [review]
Super glad we've finally been able to resolve that pesky as hell loading behaviour :tada: 
Before I spam pictures of how stuff works (because it does!), I just did another code sweep and found some potential improvements regarding cache invalidation.
In my eyes, it seems like we invalidate the domain balance query for all permissions actions, multisig and voting reputation motions that get finalized, should we limit that by type? Maybe do it in the sagas/form `onSuccess` handlers for each form.

**Review by rdig — APPROVED** [review]
VERY NICE !!!!!!!1

I cannot replicate that loading issue anymore.

Very nice work for everyone involved in this fix :100: 


https://github.com/user-attachments/assets/86dc4292-3a18-41bb-ad07-c4b9a6f9f6fc

**Review by bassgeta — APPROVED** [review]
LET'S GO AND GET THIS DASHBOARD LIVE :tada: 
So all the previous steps still work, I'm not gonna drop all of them in here again, so I'll paste in a few additional tests I've added.
1. An ordinary simple payment action works :heavy_check_mark: 
image
image
2. Switching between teams real fast doesn't just keep infinitely loading :100: 
https://github.com/user-attachments/assets/da48b542-650f-4ea6-9422-cd11d77fb43e
3. Voting reputation motions change the numbers :heavy_check_mark: 
image
image
image
image

4. Multi-sig motions work :heavy_check_mark: 
image
image
image

5. Opening up finalized motions doesn't trigger invalidation

https://github.com/user-attachments/assets/85569206-a55a-48f8-892c-b4051b32cbcb

6. Advanced payments refresh on the payment step, opening up that action again doesn't trigger a refetch :100: 

https://github.com/user-attachments/assets/6f8328d0-fff7-4cac-9171-b1f0455012f7


Great work on this one, the battle scars area real :ship:
